finagle/h2/src/main/scala/io/netty/handler/codec/http2/H2FrameCodec.scala

```diff
@@ -210,78 +207,73 @@ object H2FrameCodec {
       val fr = new Http2InboundFrameLogger(new DefaultHttp2FrameReader, frameLogger)
       new DefaultHttp2ConnectionDecoder(conn, encoder, fr)
     }
-    decoder.frameListener(new FrameListener)
 
-    val handler = new ConnectionHandler(decoder, encoder, settings)
-    new H2FrameCodec(handler)
+    val handler = new ConnectionHandler(decoder, encoder, settings, streamKey)
+    new H2FrameCodec(handler, streamKey)
   }
 
   private[this] lazy val frameLogger = new Http2FrameLogger(io.netty.handler.logging.LogLevel.TRACE, getClass)
 
   private class ConnectionHandler(
     decoder: Http2ConnectionDecoder,
     encoder: Http2ConnectionEncoder,
-    initialSettings: Http2Settings
-  ) extends Http2ConnectionHandler(decoder, encoder, initialSettings) {
-
-    override protected def onStreamError(
-      ctx: ChannelHandlerContext,
-      cause: Throwable,
-      exc: Http2Exception.StreamException
-    ): Unit =
-      try {
-        if (connection.stream(exc.streamId) != null) {
-          ctx.fireExceptionCaught(exc); ()
-        }
-      } finally {
-        super.onStreamError(ctx, cause, exc)
-      }
-  }
+    initialSettings: Http2Settings,
+    streamKey: PropertyKey
+  ) extends Http2FrameCodec(encoder, decoder, initialSettings) {
 
-  private class FrameListener extends Http2FrameAdapter {
+    decoder.frameListener(new FrameListener(connection, streamKey))
 
-    override def onRstStreamRead(ctx: ChannelHandlerContext, id: Int, code: Long): Unit = {
-      val rst = new DefaultHttp2ResetFrame(code)
-      rst.streamId(id)
-      ctx.fireChannelRead(rst); ()
-    }
+    private class FrameListener(connection: Http2Connection, streamKey: PropertyKey) extends Http2FrameAdapter {
 
-    override def onHeadersRead(
-      ctx: ChannelHandlerContext,
-      streamId: Int,
-      headers: Http2Headers,
-      streamDependency: Int,
-      weight: Short,
-      exclusive: Boolean,
-      padding: Int,
-      eos: Boolean
-    ): Unit = onHeadersRead(ctx, streamId, headers, padding, eos)
-
-    override def onHeadersRead(
-      ctx: ChannelHandlerContext,
-      streamId: Int,
-      headers: Http2Headers,
-      padding: Int,
-      eos: Boolean
-    ): Unit = {
-      val hdrs = new DefaultHttp2HeadersFrame(headers, eos, padding)
-      hdrs.streamId(streamId)
-      ctx.fireChannelRead(hdrs); ()
-    }
+      val key = streamKey
 
-    override def onDataRead(
-      ctx: ChannelHandlerContext,
-      streamId: Int,
-      content: ByteBuf,
-      padding: Int,
-      eos: Boolean
-    ): Int = {
-      val data = new DefaultHttp2DataFrame(content.retain(), eos, padding)
-      data.streamId(streamId)
-      ctx.fireChannelRead(data)
-      0 // bytes are marked as consumed via WindowUpdateFrame writes
-    }
+      def requireStream(streamId: Int): Http2FrameStream = {
+        val stream = connection.stream(streamId).getProperty(key).asInstanceOf[Http2FrameStream] match {
+          case null =>
```

Comment:
TIOLI: rather than 
```scala
match somethingThatCanBeNull {
  case null => ...
  case thing => thing
}
```
you could also say
```scala
Option(somethingThatCanBeNull).getOrElse(...)
```

Reply:
my personal vote is to LI since this way is more explicit rather than delegating to Option's null handling.